src/env.cc

```diff
@@ -385,9 +388,23 @@ Environment::Environment(IsolateData* isolate_data,
   CreateProperties();
 }
 
-CompileFnEntry::CompileFnEntry(Environment* env, uint32_t id)
-    : env(env), id(id) {
-  env->compile_fn_entries.insert(this);
+static void WeakCallbackCompiledFn(
+    const WeakCallbackInfo<CompiledFnEntry>& data) {
+  CompiledFnEntry* entry = data.GetParameter();
+  entry->env->id_to_function_map.erase(entry->id);
+  delete entry;
+}
+
+CompiledFnEntry::CompiledFnEntry(Environment* env,
+                                 uint32_t id,
+                                 Local<Function> function)
+    : env(env),
+      id(id),
+      function(env->isolate(), function),
```

Comment:
We don’t need to store the function here, right?

Reply:
we were already storing this function here before, just not inside a struct. I added the struct because it also needs to hold the weak ref to the unbound script

Reply:
I know, but we still don’t need the function (and are just unnecessarily keeping it alive this way), right? The previous usage was because we tried to wait for GC on the function itself, but that’s gone now…

Reply:
This struct is held in the id_to_function map. That map needs to be able to go from id to function, the module resolution system uses it.

Reply:
I see, missed that the function is actually used as the key for looking up the dynamic import callback in JS land … I feel like there should be a better way but I think that’s also out of scope for this PR :+1: